### ISSUE TITLE
no-undef errors incorrectly if function is defined within block and block binding is enabled

### ISSUE THREAD

**Issue by bgw** [eslint/eslint#3130]
```js
var foo = true;

if (foo) {
  function bar() { console.log('hey'); }
}

if (foo) {
  bar();
}
```

gives

![screen shot 2015-07-22 at 11 50 43 am](https://cloud.githubusercontent.com/assets/180404/8834246/f0911ff8-3067-11e5-9d27-a7ec41c2de44.png)

When using espree with `blockBindings: true` or babel-eslint.

```yaml
---
parser: espree
ecmaFeatures:
  blockBindings: true
rules:
  no-undef: 2
env:
  node: true
```

It seems related to [this bit of code in escope](https://github.com/estools/escope/blob/dd36f30eb8b23bae1bc6ff58bf086349b954c1df/src/referencer.js#L587).

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by btmills** [comment]
In ES5, function declarations inside blocks were invalid, so I'm assuming this is ES6, in which case `bar` would only be defined within the consequent of the first `if`, right? If you combine the contents of the two `if`s, does it still report an error?

**Comment by DmitrySoshnikov** [comment]
So in ES6 it should be be "bar is not defined", but in ES5 it should throw error "function declarations cannot be defined inside blocks" (even if most of the browsers deviated from the spec, and implemented the function statements in blocks -- though, with a lot of bugs and deviations even from each other).

**Comment by btmills** [comment]
> So in ES6 it should be be "bar is not defined"

Yep

> in ES5 it should throw error "function declarations cannot be defined inside blocks"

Yes, though we have a separate rule for this: [no-inner-declarations](http://eslint.org/docs/rules/no-inner-declarations)

**Comment by bgw** [comment]
@btmills @DmitrySoshnikov, Thanks, it seems like this isn't a bug in eslint, just a strange detail of the spec.

**Comment by michaelficarra** [comment]
The tricky part here is that if you wrap the entire thing in an IIFE

```js
(function(){
  var foo = true;

  if (foo) {
    function bar() { console.log('hey'); }
  }

  if (foo) {
    bar();
  }
}());
```

then this program should not have any runtime errors according to ES6. Block-level function declarations leak their binding to the function scope only if 

1. not in strict mode
1. doing so would not cause an early error (due to a conflicting block binding in the containing scope or anywhere along the way)
1. the binding would not be hoisted to the top level
